lib/web_ui/flutter_js/src/instantiate_wasm.js

```diff
@@ -0,0 +1,17 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This is a little helper function that helps us start the fetch and compilation
+// of an emscripten wasm module in parallel with the fetch of its script.
+export const createWasmInstantiator = (url) => {
+  const modulePromise = WebAssembly.compileStreaming(fetch(url));
+  return (imports, successCallback) => {
```

Comment:
This `successCallback` here is a little bit weird in an async function, shouldn't you be able to:

```js
const modulePromise = WebAssembly.compileStreaming(fetch(url));
return (imports) => new Promise((resolve, reject) => {
  modulePromise.then((module) => {
    WebAssembly.instantiate(module, imports).then((instance) => {
      resolve({
        instance: instance,
        module: module,
      });
    });
  });
});
```

And from the outside you can:

```js
const { module, instance } = await wasmInstantiator(url)(imports); // Very Functional!
```

(I'm sure this can be simplified even further but my 🧠 is 🍳)

Reply:
(I wonder if the promise resolver can be an async function so it can be implemented with `await` rather than `.then`)

Reply:
It certainly would be nice if this worked the way you're describing, but this is meant to be plugged into an emscripten API (`instantiateWasm`) which doesn't work with promises. I wish they did, I'm not really sure why their API isn't promise-based, but the (admittedly very strange) contract of the API is that the function takes a `successCallback`, and you can either synchronously return the module instance, or return an empty JS object (`{}`) and then call `successCallback` later asynchronously with the instance and module. Weird, I know, but it's outside of our control.

See https://emscripten.org/docs/api_reference/module.html#Module.instantiateWasm

Reply:
Awwwgh okay!